lib/Reader/readerir.cpp

```diff
@@ -3783,6 +3797,253 @@ IRNode *GenIR::genArrayElemAddress(IRNode *Array, IRNode *Index,
+  // (doesn't count toward the rank).
+  *Rank = IsStore ? Sig->numArgs - 1 : Sig->numArgs;
+
+  // Early out if we don't want to handle it.
+  // NOTE: rank 1 requires special handling (not done here).
+  if (*Rank > ArrayIntrinMaxRank || *Rank <= 1) {
+    return false;
+  }
+
+  // Figure out the element type
+  CorInfoType CorType = CORINFO_TYPE_UNDEF;
+  CORINFO_CLASS_HANDLE Class = nullptr;
+  uint32_t ElemSize = 0;
+  if (IsStore) {
+    CORINFO_ARG_LIST_HANDLE ArgLst = Sig->args;
+    // Skip arguments for each dimension.
+    for (uint32_t R = 0; R < *Rank; R++) {
+      ArgLst = getArgNext(ArgLst);
+    }
+    CORINFO_CLASS_HANDLE ArgType;
+    CorType = strip(getArgType(Sig, ArgLst, &ArgType));
+
+    ASSERTNR(CorType != CORINFO_TYPE_VAR); // common generics trouble
+
+    if (CorType == CORINFO_TYPE_CLASS || CorType == CORINFO_TYPE_VALUECLASS) {
+      Class = getArgClass(Sig, ArgLst);
+    } else if (CorType == CORINFO_TYPE_REFANY) {
+      Class = getBuiltinClass(CLASSID_TYPED_BYREF);
+    } else {
+      Class = nullptr;
+    }
+  } else {
+    CorType = Sig->retType;
+    Class = Sig->retTypeClass;
+  }
+
+  // Bail if the element type is a value class.
+  if (CorType == CORINFO_TYPE_VALUECLASS) {
+    return false;
+  }
+
+  if (IsStore || IsLoadAddr) {
+    CORINFO_CLASS_HANDLE GCClass = Class;
+    CorInfoType GCCorType = CorType;
+
+    // If it is BYREF then we need to find the child type
+    // for the next security check.
+    if (CorType == CORINFO_TYPE_BYREF) {
+      CORINFO_CLASS_HANDLE ClassTmp = nullptr;
+      GCCorType = getChildType(Class, &ClassTmp);
+      GCClass = ClassTmp;
+    }
+
+    // See if the element is an object.
+    if (GCCorType == CORINFO_TYPE_STRING || GCCorType == CORINFO_TYPE_CLASS) {
+      // If it's not final, we can't expand due to security reasons.
+      uint32_t ClassAttribs = getClassAttribs(GCClass);
+      if (!(ClassAttribs & CORINFO_FLG_FINAL)) {
+        return false;
+      }
+    }
+  }
+
+  *ElemCorType = CorType;
+  *ElemType = getType(CorType, Class);
+
+  return true;
+}
+
+IRNode *GenIR::mdArrayRefAddr(uint32_t Rank, llvm::Type *ElemType) {
+  IRNode *Indices[ArrayIntrinMaxRank];
+  for (uint32_t I = Rank; I > 0; --I) {
+    Indices[I - 1] = ReaderOperandStack->pop();
+  }
+
+  IRNode *Array = ReaderOperandStack->pop();
+
+  // The memory layout of the array is as follows:
+  //     PTR          MethodTable for ArrayType
+  //     unsigned     length      Total number of elements in the array
+  //     unsigned     alignpad  for 64 bit alignment
+  //     unsigned     dimLengths[rank]
+  //     unsigned     dimBounds[rank]
+  //     Data elements follow this.
+  //
+  // Note that this layout is somewhat described in corinfo.h,
+  // CORINFO_Array and CORINFO_RefArray.
+  //
+  // Valid indexes are
+  //     dimBounds[i] <= index[i] < dimBounds[i] + dimLengths[i]
+
+  assert(!ElemType->isStructTy());
+
+  if (UseExplicitNullChecks) {
+    Array = genNullCheck(Array);
+  }
```

Comment:
It would be good to have a comment saying that if UseExplicitNullChecks were false, we'd need to annotate the first load as exception-bearing.